staging/src/k8s.io/apiserver/pkg/registry/rest/delete.go

```diff
@@ -103,20 +103,23 @@ func BeforeDelete(strategy RESTDeleteStrategy, ctx context.Context, obj runtime.
 		// 2. Delete the object from storage.
 		// If the update succeeds, but the delete fails (network error, internal storage error, etc.),
 		// a resource was previously left in a state that was non-recoverable.  We
-		// check if the existing stored resource has a grace period as 0 and if so
+		// check if the existing stored resource has a grace period <= 0 and if so
```

Comment:
This is incorrect and not the desired behavior.  metav1.DeleteOptions.GracePeriodSeconds is allowed to be negative.  This should be interpreted as "use pod's value" and should be handled differently than GracePeriodSeconds in the podspec.

Reply:
> This should be interpreted as "use pod's value"

Where is that defined? I'm not seeing anything that passed -1 to the API to meant that

Reply:
Looks like you are correct: https://github.com/kubernetes/apimachinery/blob/v0.20.4/pkg/apis/meta/v1/types.go#L492

According to the API, it must be non-negative, so we should reject any call with a negative value rather than overwrite it.

Reply:
> According to the API, it must be non-negative, so we should reject any call with a negative value rather than overwrite it.

Accepting calls and trimming to 0 is more compatible. If there are clients issuing deletes with a -1, actually deleting the resource is closer to the intent than letting it hang forever or requiring an update to the API client.

Reply:
If the API says non-negative only, we should reject, otherwise the API documentation is wrong, at the very least we should update it.

As I mentioned in my other comment, negative values are probably an indication that someone want's the pod's default value rather than a 'delete now' value.  If they wanted to delete now, they would set a 0 or a 1.